Release proposal: v4.2.0 'Argon' (LTS)

jasnell opened the pull request:
Aiming for Monday release. This is for the LTS release. Notable changes still need to be documented and there's still likely a small handful of issues that need to be closed and included (see https://github.com/nodejs/node/pulls?page=1&q=is%3Apr+label%3Aland-on-v4.x+-label%3Alts-landed-on-v4.x&utf8=%E2%9C%93)

---

* [[`8383c4fe00`](https://github.com/nodejs/node/commit/8383c4fe00)] - **assert**: support arrow functions in .throws() (Ben Noordhuis) [#3276](https://github.com/nodejs/node/pull/3276)
* [[`3eaa593a32`](https://github.com/nodejs/node/commit/3eaa593a32)] - **async_wrap**: correctly pass parent to init callback (Trevor Norris) [#3216](https://github.com/nodejs/node/pull/3216)
* [[`54795620f6`](https://github.com/nodejs/node/commit/54795620f6)] - **buffer**: don't abort on prototype getters (Trevor Norris) [#3302](https://github.com/nodejs/node/pull/3302)
* [[`660f7591c8`](https://github.com/nodejs/node/commit/660f7591c8)] - **buffer**: FreeCallback should be tied to ArrayBuffer (Fedor Indutny) [#3198](https://github.com/nodejs/node/pull/3198)
* [[`651a5b51eb`](https://github.com/nodejs/node/commit/651a5b51eb)] - **buffer**: only check if instance is Uint8Array (Trevor Norris) [#3080](https://github.com/nodejs/node/pull/3080)
* [[`d5a1b1ad7c`](https://github.com/nodejs/node/commit/d5a1b1ad7c)] - **buffer**: clean up usage of __proto__ (Trevor Norris) [#3080](https://github.com/nodejs/node/pull/3080)
* [[`af24376e18`](https://github.com/nodejs/node/commit/af24376e18)] - **build**: Intl: deps: bump ICU to 56.1 (GA) (Steven R. Loomis) [#3281](https://github.com/nodejs/node/pull/3281)
* [[`9136359d57`](https://github.com/nodejs/node/commit/9136359d57)] - **build**: make icu download path customizable (Johan Bergström) [#3200](https://github.com/nodejs/node/pull/3200)
* [[`b3c5ad10a8`](https://github.com/nodejs/node/commit/b3c5ad10a8)] - **build**: add --with-arm-fpu option (Jérémy Lal) [#3228](https://github.com/nodejs/node/pull/3228)
* [[`f00f3268e4`](https://github.com/nodejs/node/commit/f00f3268e4)] - **build**: intl: avoid 'duplicate main()' on ICU 56 (Steven R. Loomis) [#3066](https://github.com/nodejs/node/pull/3066)
* [[`071c72a6a3`](https://github.com/nodejs/node/commit/071c72a6a3)] - **deps**: upgrade to npm 2.14.7 (Kat Marchán) [#3299](https://github.com/nodejs/node/pull/3299)
* [[`8b50e95f06`](https://github.com/nodejs/node/commit/8b50e95f06)] - **(SEMVER-MINOR)** **deps**: backport 1ee712a from V8 upstream (Julien Gilli) [#3036](https://github.com/nodejs/node/pull/3036)
* [[`40a159e4f4`](https://github.com/nodejs/node/commit/40a159e4f4)] - **doc**: Amend capitalization of word JavaScript (Dave Hodder) [#3285](https://github.com/nodejs/node/pull/3285)
* [[`6dd34761fd`](https://github.com/nodejs/node/commit/6dd34761fd)] - **doc**: add method links in dns.markdown (Alejandro Oviedo) [#3196](https://github.com/nodejs/node/pull/3196)
* [[`333e8336be`](https://github.com/nodejs/node/commit/333e8336be)] - **doc**: add method links in child_process.markdown (Alejandro Oviedo) [#3186](https://github.com/nodejs/node/pull/3186)
* [[`0cfc6d39ca`](https://github.com/nodejs/node/commit/0cfc6d39ca)] - **doc**: recommend Infinity on emitter.setMaxListeners (Jason Karns) [#2559](https://github.com/nodejs/node/pull/2559)
* [[`d4fc6d93ef`](https://github.com/nodejs/node/commit/d4fc6d93ef)] - **doc**: add help repo link to CONTRIBUTING.md (Doug Shamoo) [#3233](https://github.com/nodejs/node/pull/3233)
* [[`28aac7f19d`](https://github.com/nodejs/node/commit/28aac7f19d)] - **doc**: add TLS session resumption example (Roman Reiss) [#3147](https://github.com/nodejs/node/pull/3147)
* [[`365cf22cce`](https://github.com/nodejs/node/commit/365cf22cce)] - **doc**: update AUTHORS list (Rod Vagg) [#3211](https://github.com/nodejs/node/pull/3211)
* [[`d4399613b7`](https://github.com/nodejs/node/commit/d4399613b7)] - **doc**: standardize references to userland (Martial) [#3192](https://github.com/nodejs/node/pull/3192)
* [[`75de258376`](https://github.com/nodejs/node/commit/75de258376)] - **doc**: fix spelling in Buffer documentation (Rod Machen) [#3226](https://github.com/nodejs/node/pull/3226)
* [[`725c7276dd`](https://github.com/nodejs/node/commit/725c7276dd)] - **doc**: fix README.md link to joyent/node intl wiki (Steven R. Loomis) [#3067](https://github.com/nodejs/node/pull/3067)
* [[`4a35ba4966`](https://github.com/nodejs/node/commit/4a35ba4966)] - **(SEMVER-MINOR)** **fs**: include filename in watch errors (charlierudolph) [#2748](https://github.com/nodejs/node/pull/2748)
* [[`2ddbbfd164`](https://github.com/nodejs/node/commit/2ddbbfd164)] - **http**: cork/uncork before flushing pipelined res (Fedor Indutny) [#3172](https://github.com/nodejs/node/pull/3172)
* [[`f638402e2f`](https://github.com/nodejs/node/commit/f638402e2f)] - **http**: add comment about `outputSize` in res/server (Fedor Indutny) [#3128](https://github.com/nodejs/node/pull/3128)
* [[`1850879b0e`](https://github.com/nodejs/node/commit/1850879b0e)] - **js_stream**: prevent abort if isalive doesn't exist (Trevor Norris) [#3282](https://github.com/nodejs/node/pull/3282)
* [[`63644dd1cd`](https://github.com/nodejs/node/commit/63644dd1cd)] - **lib**: remove redundant code, add tests in timers.js (Rich Trott) [#3143](https://github.com/nodejs/node/pull/3143)
* [[`74f443583c`](https://github.com/nodejs/node/commit/74f443583c)] - **module**: use UNC paths when loading native addons (Justin Chase) [#2965](https://github.com/nodejs/node/pull/2965)
* [[`9d65528b01`](https://github.com/nodejs/node/commit/9d65528b01)] - **(SEMVER-MINOR)** **node**: add -c|--check CLI arg to syntax check script (Dave Eddy) [#2411](https://github.com/nodejs/node/pull/2411)
* [[`589287b2e3`](https://github.com/nodejs/node/commit/589287b2e3)] - **src**: convert BE-utf16-string to LE before search (Karl Skomski) [#3295](https://github.com/nodejs/node/pull/3295)
* [[`2314378f06`](https://github.com/nodejs/node/commit/2314378f06)] - **src**: fix u-a-free if uv returns err in ASYNC_CALL (Karl Skomski) [#3049](https://github.com/nodejs/node/pull/3049)
* [[`d99336a391`](https://github.com/nodejs/node/commit/d99336a391)] - **(SEMVER-MINOR)** **src**: replace naive search in Buffer::IndexOf (Karl Skomski) [#2539](https://github.com/nodejs/node/pull/2539)
* [[`546e8333ba`](https://github.com/nodejs/node/commit/546e8333ba)] - **(SEMVER-MINOR)** **src**: fix --abort-on-uncaught-exception (Jeremy Whitlock) [#3036](https://github.com/nodejs/node/pull/3036)
* [[`7271cb047c`](https://github.com/nodejs/node/commit/7271cb047c)] - **(SEMVER-MINOR)** **src**: add process.versions.icu (Evan Lucas) [#3102](https://github.com/nodejs/node/pull/3102)
* [[`f0f8afd879`](https://github.com/nodejs/node/commit/f0f8afd879)] - **test**: remove common.inspect() (Rich Trott) [#3257](https://github.com/nodejs/node/pull/3257)
* [[`5ca4f6f8bd`](https://github.com/nodejs/node/commit/5ca4f6f8bd)] - **test**: test `util` rather than `common` (Rich Trott) [#3256](https://github.com/nodejs/node/pull/3256)
* [[`7a5ae34345`](https://github.com/nodejs/node/commit/7a5ae34345)] - **test**: refresh temp directory when using pipe (Rich Trott) [#3231](https://github.com/nodejs/node/pull/3231)
* [[`7c85557ef0`](https://github.com/nodejs/node/commit/7c85557ef0)] - **test**: Fix test-fs-read-stream-fd-leak race cond (Junliang Yan) [#3218](https://github.com/nodejs/node/pull/3218)
* [[`26a7ec6960`](https://github.com/nodejs/node/commit/26a7ec6960)] - **test**: fix losing original env vars issue (Junliang Yan) [#3190](https://github.com/nodejs/node/pull/3190)
* [[`e922716192`](https://github.com/nodejs/node/commit/e922716192)] - **test**: remove deprecated error logging (Rich Trott) [#3079](https://github.com/nodejs/node/pull/3079)
* [[`8f29d95a8c`](https://github.com/nodejs/node/commit/8f29d95a8c)] - **test**: report timeout in TapReporter (Karl Skomski) [#2647](https://github.com/nodejs/node/pull/2647)
* [[`2d0fe4c657`](https://github.com/nodejs/node/commit/2d0fe4c657)] - **test**: linting for buffer-free-callback test (Rich Trott) [#3230](https://github.com/nodejs/node/pull/3230)
* [[`70c9e4337e`](https://github.com/nodejs/node/commit/70c9e4337e)] - **test**: make common.js mandatory via linting rule (Rich Trott) [#3157](https://github.com/nodejs/node/pull/3157)
* [[`b7179562aa`](https://github.com/nodejs/node/commit/b7179562aa)] - **test**: load common.js in all tests (Rich Trott) [#3157](https://github.com/nodejs/node/pull/3157)
* [[`bab555a1c1`](https://github.com/nodejs/node/commit/bab555a1c1)] - **test**: speed up stringbytes-external test (Evan Lucas) [#3005](https://github.com/nodejs/node/pull/3005)
* [[`ddf258376d`](https://github.com/nodejs/node/commit/ddf258376d)] - **test**: use normalize() for unicode paths (Roman Reiss) [#3007](https://github.com/nodejs/node/pull/3007)
* [[`46876d519c`](https://github.com/nodejs/node/commit/46876d519c)] - **test**: remove arguments.callee usage (Roman Reiss) [#3167](https://github.com/nodejs/node/pull/3167)
* [[`9c2748bad1`](https://github.com/nodejs/node/commit/9c2748bad1)] - **tools**: remove leftover license boilerplate (Nathan Rajlich) [#3225](https://github.com/nodejs/node/pull/3225)
* [[`5d9f83ff2a`](https://github.com/nodejs/node/commit/5d9f83ff2a)] - **tools**: apply linting to custom rules code (Rich Trott) [#3195](https://github.com/nodejs/node/pull/3195)
* [[`18a8b2ec73`](https://github.com/nodejs/node/commit/18a8b2ec73)] - **tools**: remove unused gflags module (Ben Noordhuis) [#3220](https://github.com/nodejs/node/pull/3220)
* [[`e0fffca836`](https://github.com/nodejs/node/commit/e0fffca836)] - **util**: fix for inspecting promises (Evan Lucas) [#3221](https://github.com/nodejs/node/pull/3221)
* [[`8dfdee3733`](https://github.com/nodejs/node/commit/8dfdee3733)] - **util**: correctly inspect Map/Set Iterators (Evan Lucas) [#3119](https://github.com/nodejs/node/pull/3119)
* [[`b5c51fdba0`](https://github.com/nodejs/node/commit/b5c51fdba0)] - **util**: fix check for Array constructor (Evan Lucas) [#3119](https://github.com/nodejs/node/pull/3119)

jasnell commented:
@nodejs/tsc @nodejs/lts

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/451/

jasnell commented:
@srl295 ... I know you are looking to get ICU bumped up to 56 in v4.x. Is it possible to get that work completed and landed in master by monday? If yes, I can cherry pick it into this for 4.2

srl295 commented:
This morning

El miércoles, 7 de octubre de 2015, James M Snell <notifications@github.com>
escribió:

> @srl295 <https://github.com/srl295> ... I know you are looking to get ICU
> bumped up to 56 in v4.x. Is it possible to get that work completed and
> landed in master by monday? If yes, I can cherry pick it into this for 4.2
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/3258#issuecomment-146428974>.
>

jasnell commented:
Picked three additional commits from master onto v4.x, rebased and updated the CHANGELOG

jasnell commented:
Outstanding Issues before we can close on this:

* [ ] https://github.com/nodejs/node/issues/3283

indutny commented:
Does it include my Buffer::New fix? Cc @trevnorris

On Thursday, October 8, 2015, James M Snell <notifications@github.com>
wrote:

> Outstanding Issues before we can close on this:
>
>    - #3283 <https://github.com/nodejs/node/issues/3283>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/3258#issuecomment-146638776>.
>

indutny commented:
It does, thanks!

jasnell commented:
@nodejs/collaborators ... this is a last call for v4.2! I can get semver-patch and conservative semver-minor commits that land in master pulled over into v4.x but want to get things locked down so we can run a variety of tests over the weekend to make sure things are solid before cutting the v4.2.0 LTS release on monday. If there are any PRs you think should be landed in v4.x, attach the land-on-v4.x label to the PR and mention me in a comment.

jasnell commented:
Pulled in a couple more commits, rebased and updated changelog. Still waiting on the ICU 56 bump and any other last call items

jasnell commented:
ICU update landed. Now waiting for a fix for #3283 .

ahdinosaur commented:
> v4.2.0

:yum:

Fishrock123 commented:
@jasnell you still need https://github.com/nodejs/node/pull/3212

jasnell commented:
Yep. This makes the assumption that that will close by then.
On Oct 8, 2015 7:36 PM, "Jeremiah Senkpiel" <notifications@github.com>
wrote:

> @jasnell <https://github.com/jasnell> you still need #3212
> <https://github.com/nodejs/node/pull/3212>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/3258#issuecomment-146739604>.
>

rvagg commented:
@jasnell we have a conflict with my PR @ https://github.com/nodejs/node/pull/3212 for adding `process.release.lts`. I'm proposing a simple version that just adds it as a straight property without the node_version.h stuff—I don't see a reason this change needs to exist at all on master which is why there's no special macro work. At a minimum I'd like to include the test and the doc changes from #3212, I'm also in favour of "Argon" vs "argon".

Thoughts?

jasnell commented:
@rvagg ... I can easily back my changes on that out of the proposal PR. I
do think, however, that the macro approach is better in the long run given
that it would be less we need to remember to do in future LTS releases but
I'm not overly concerned about that.
On Oct 9, 2015 4:39 AM, "Rod Vagg" <notifications@github.com> wrote:

> @jasnell <https://github.com/jasnell> we have a conflict with my PR @
> #3212 <https://github.com/nodejs/node/pull/3212> for adding
> process.release.lts. I'm proposing a simple version that just adds it as
> a straight property without the node_version.h stuff—I don't see a reason
> this change needs to exist at all on master which is why there's no special
> macro work. At a minimum I'd like to include the test and the doc changes
> from #3212 <https://github.com/nodejs/node/pull/3212>, I'm also in favour
> of "Argon" vs "argon".
>
> Thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/3258#issuecomment-146840337>.
>

Fishrock123 commented:
> "Argon"

+1 this spelling

zkat commented:
Can we include #3299 in this, too?

(also, kinda wish we'd call it "Blaze It" but oh well. That :ship: has sailed)

jasnell commented:
* Landed a few more commits on v4.x, including the npm version bump.
* Rebased the PR. 
* Changed `argon` to `Argon` -- still need to reconcile with @rvagg's #3212 

Locking this down. Not going to pull anything else from master unless it's a bug fix. 
Need a resolution for https://github.com/nodejs/node/issues/3303 
Other than that, things are looking solid.

@nodejs/collaborators @nodejs/lts ... please start smoke testing this!

jasnell commented:
@rvagg @Fishrock123 @mhdawson @nodejs/lts ... anything further needed on this? Looking to cut the release on Monday morning.

Fishrock123 commented:
@jasnell https://github.com/nodejs/node/commit/68990948fe473262ec5f7d0b06112835ffd613aa should probably go in. It fixes https://github.com/nodejs/node/issues/2323.

rvagg commented:
Suggestion for notable items section changes from what you currently have, to be consistent:

```
## Notable changes

The first Node.js LTS release! See https://github.com/nodejs/LTS/ for details of the LTS process.

* **icu**: updated to version 56 with significant performance improvements (Steven R. Loomis) [#3281](https://github.com/nodejs/node/pull/3281)
* **node**: Added new `-c` (or `--check`) command line argument for checking script syntax without executing the code (Dave Eddy) [#2411](https://github.com/nodejs/node/pull/2411)
* **npm**: Upgraded to npm 2.14.7 from 2.14.4, see [release notes](https://github.com/npm/npm/releases/tag/v2.14.7) for full details (Kat Marchán) [#3299](https://github.com/nodejs/node/pull/3299)
* **process**:
  - Added `process.versions.icu` to hold the current ICU library version (Evan Lucas) [#3102](https://github.com/nodejs/node/pull/3102)
  - Added `process.release.lts` to hold the current LTS codename when the binary is from an active LTS release line (Rod Vagg) [#3212](https://github.com/nodejs/node/pull/3212)
```

Also, `Known issues` instead of `Known Issues`.

rvagg commented:
@jasnell I think https://github.com/nodejs/node/pull/3212 is good to go and matches what you have here, it hasn't had review since I changed it, however, so can I get you to merge it if you start working on the release soon? I'll be :zzz: for the next 8 hours or so, so don't count on me being around.

rvagg commented:
@jasnell also, when you post to the nodejs.org blog, perhaps put a little blurb at the top about what it means for this to be LTS, where to get more information and to expect a v5 that's _not_ LTS to be out by the end of the month. I'm betting there's a significant number of people that haven't heard the message about LTS (and a lot who have but don't know what LTS stands for so have been ignoring it).

jasnell commented:
@rvagg excellent. Ok. I'll be starting on finalizing this around 8:00am
Pacific.
On Oct 12, 2015 5:21 AM, "Rod Vagg" <notifications@github.com> wrote:

> @jasnell <https://github.com/jasnell> I think #3212
> <https://github.com/nodejs/node/pull/3212> is good to go and matches what
> you have here, it hasn't had review since I changed it, however, so can I
> get you to merge it if you start working on the release soon? I'll be [image:
> :zzz:] for the next 8 hours or so, so don't count on me being around.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/3258#issuecomment-147379822>.
>

Fishrock123 commented:
@jasnell one last one: a7130245796dc6a422063af2b0b55ca1120a9dc9, if this is still open.

jasnell commented:
@Fishrock123 ... I just finished building the main release artifacts and was smoketesting a bit before tagging. Is this one critical?

jasnell commented:
nm... I'll go ahead and land it. Needs to be the last one tho. Will kick off a new iojs+release run in a sec

indutny commented:
@jasnell maybe 51325c03d42ec10bf8b52c3555b0bfbf0894d0be too?

jasnell commented:
@indutny @Fishrock123 ... Y'all are killing me here ;-P

jasnell commented:
Release done.

Trott commented:
![YESSSSS](http://i.imgur.com/JKTF9GJ.gif)

jasnell commented:
I'm gonna go lay down now. I need a nap.

skomski commented:
@jasnell Sorry didn't see that thread earlier but it wouldn't it make sense to add `src: replace naive search in Buffer::IndexOf` under notable changes? And the additional argument for `indexOf` needs to be documented. I was a bit surprised this PR was merged first into LTS.

jasnell commented:
@skomski ... PRs to add it to the notable changes in the CHANGELOG for both master, v4.x and the new.nodejs.org entry would be good. You're right, it should be called out. Landing semver-minor bits like that is fine as we come up to the LTS release.. landing it *after* we cut the LTS is a no-no.. and I felt it was a good enough improvement that it made sense to get it in (especially since we were doing a minor bump for the LTS anyway).
